// handsontable/3rdparty/walkontable/src/index.js
"use strict";

exports.__esModule = true;
var _calculator = require("./calculator");
exports.ViewportColumnsCalculator = _calculator.ViewportColumnsCalculator;
exports.ViewportRowsCalculator = _calculator.ViewportRowsCalculator;
exports.DEFAULT_COLUMN_WIDTH = _calculator.DEFAULT_COLUMN_WIDTH;
var _coords = _interopRequireDefault(require("./cell/coords"));
exports.CellCoords = _coords.default;
var _range = _interopRequireDefault(require("./cell/range"));
exports.CellRange = _range.default;
var _core = _interopRequireDefault(require("./facade/core"));
exports.default = _core.default;
exports.Core = _core.default;
var _selection = require("./selection");
exports.Selection = _selection.Selection;
exports.HIGHLIGHT_ACTIVE_HEADER_TYPE = _selection.ACTIVE_HEADER_TYPE;
exports.HIGHLIGHT_AREA_TYPE = _selection.AREA_TYPE;
exports.HIGHLIGHT_FOCUS_TYPE = _selection.FOCUS_TYPE;
exports.HIGHLIGHT_FILL_TYPE = _selection.FILL_TYPE;
exports.HIGHLIGHT_HEADER_TYPE = _selection.HEADER_TYPE;
exports.HIGHLIGHT_ROW_TYPE = _selection.ROW_TYPE;
exports.HIGHLIGHT_COLUMN_TYPE = _selection.COLUMN_TYPE;
exports.HIGHLIGHT_CUSTOM_SELECTION_TYPE = _selection.CUSTOM_SELECTION_TYPE;
var Renderer = _interopRequireWildcard(require("./renderer"));
exports.Renderer = Renderer;
var _orderView = require("./utils/orderView");
exports.OrderView = _orderView.OrderView;
exports.SharedOrderView = _orderView.SharedOrderView;
var _nodesPool = require("./utils/nodesPool");
exports.NodesPool = _nodesPool.NodesPool;
var _eventManager = require("../../../eventManager");
exports.getListenersCounter = _eventManager.getListenersCounter;
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
